relayer/src/channel/version.rs

```diff
@@ -0,0 +1,135 @@
+//! Helper module for the relayer channel logic.
+//!
+//! Provides support for resolving the appropriate
+//! channel version to be used in a channel open
+//! handshake.
+
+use tracing::{debug, info, warn};
+
+use ibc::{
+    applications::{ics20_fungible_token_transfer, ics27_interchain_accounts},
+    core::{ics04_channel::Version, ics24_host::identifier::PortId},
+};
+use ibc_proto::ibc::core::{channel::v1::Counterparty, port::v1::QueryAppVersionRequest};
+
+use crate::error::ErrorDetail;
+use crate::{
+    chain::handle::ChainHandle,
+    channel::{Channel, ChannelError},
+    error::Error,
+};
+
+/// Defines the context in which to resolve a channel version.
+///
+/// This context distinguishes between different steps of
+/// the channel open handshake protocol. Currently, we only
+/// need to distinguish between the OpenInit step and any other.
+#[derive(Debug)]
+pub enum ResolveContext {
+    ChanOpenInit,
+    Other,
+}
+
+/// Resolves the [`Version`] to use during a channel
+/// open handshake.
+///
+/// The channel version depends on the exact channel
+/// handshake step [`ResolveContext`], as well as
+/// on the current state of the channel.
+pub fn resolve<ChainA: ChainHandle, ChainB: ChainHandle>(
+    ctx: ResolveContext,
+    channel: &Channel<ChainA, ChainB>,
+) -> Result<Version, ChannelError> {
+    let port_id = channel.dst_port_id();
+
+    match ctx {
+        ResolveContext::ChanOpenInit => {
+            // When the channel is uninitialized, the relayer will use the
+            // default channel version, depending on the port id.
+            default_by_port(port_id)
+        }
+        ResolveContext::Other => {
+            // Resolve the version by querying the application version on destination chain
+            let dst_chain_id = channel.dst_chain().id();
+            info!("resolving channel version for port id='{}', context='{:?}' by retrieving dst chain '{}' app version", port_id, ctx, dst_chain_id);
+
+            // Note the compatibility logic below:
+            // The destination chain may or may not implement `QueryAppVersionRequest`,
+            // which is only available from ibc-go v2.0.0 or newer versions.
+            // We detect if it is implemented by calling `is_unimplemented_port_query`.
+            // If unimplemented, we can recover by simply using the default version.
+            // This ensures compatibility with ibc-go v2.0.0+ and pre-v2.0.0.
+            match dst_app_version(channel) {
+                Ok(v) => Ok(v),
+                Err(e) => {
+                    if let ErrorDetail::GrpcStatus(s) = e.detail() {
+                        if s.is_unimplemented_port_query() {
+                            // Ensure compatibility & recover from the error,
+                            // by using the default version.
+                            debug!("resorting to default version because dst chain '{}' does not expose app version gRPC endpoint", dst_chain_id);
```

Comment:
```suggestion
                            debug!(chain_id = %dst_chain_id, "resorting to default version because destination chain does not expose app version gRPC endpoint");
```